Adding the "Push active document" button

nGolubeva opened the pull request.

Commit:
Adding the "Push active document" button

jenstroeger committed:
Fix typos.

jenstroeger committed:
Replace “Push Active” button, hook up JSX stub functions.

gregoriopellegrino committed:
First push for Indesign to RTF script

jenstroeger commented on extensions/Bookalope/jsx/bookalope.jsx:
```diff
@@ -193,3 +193,264 @@ function bookalopeCreateDocument(idmlFileName, bookId, bookflowId, betaHost) {
+		for (var i = 0; i < g.length; i++) {
+			if (g[i].itemLink == null) {
+				image_file = File (uniqueName(outfolder+'/img', '.png'));
+				export_other (g[i], ExportFormat.PNG_FORMAT, image_file);
```
@gregoriopellegrino, I can’t find the function implementation for `export_other()` — not in this PR and not in the [InDesign ExtendScript API](https://www.indesignjs.de/extendscriptAPI/indesign-latest/) 🤔

gregoriopellegrino replied:
Fixed :), I'm trying to test it locally, but I have difficulty activating it on InDesign...

jenstroeger replied:
Thank you, @gregoriopellegrino!

> I have difficulty activating it on InDesign...

You mean to install the extension locally for InDesign? [This link](https://github.com/bookalope/InDesign-CEP#installing) doesn’t work?

jenstroeger replied:
@gregoriopellegrino I accidentally dropped your commit when I force-pushed after a rebase — could you please commit that function again?

gregoriopellegrino replied:
Done!

jenstroeger replied:
Thank you! ☺️

gregoriopellegrino committed:
Added export_other function

gregoriopellegrino committed:
Fixing bugs

gregoriopellegrino commented:
I've fixed some bugs. Now it seems ok, I receive a green status after the process, but I can't find the bookflow in Bookalope (beta).

Maybe because there is a previous upload stucked in processing: https://beta.bookalope.net/books/d7c7a29552c8483f9bbe922819136109

jenstroeger commented:
Great, thanks @gregoriopellegrino, I’ll take a look. Also, I just checked: your [Book d7c7a295](https://beta.bookalope.net/books/d7c7a29552c8483f9bbe922819136109) has one [Bookflow e750bf12](https://beta.bookalope.net/bookflows/e750bf12b9e0481ba88c5722730e0101/convert) in `convert` step, so all should be well (I didn’t check the result).

gregoriopellegrino committed:
Fixing bug with graphics without area

gregoriopellegrino committed:
Ignore .DS_store on mac

jenstroeger committed:
Minor fixes to the Extension Client code.

jenstroeger committed:
Refactor RTF code.

jenstroeger commented:
@gregoriopellegrino I refactored/restructured your code a little without changing its functionality. It’s still working for me.

Would you mind taking a look? I left a few `TODO` comments and would like to know your thoughts. Happy to open individual conversations using PR comments on the code, if that helps?

jenstroeger commented on extensions/Bookalope/jsx/bookalope.jsx:
```diff
@@ -193,3 +193,347 @@ function bookalopeCreateDocument(idmlFileName, bookId, bookflowId, betaHost) {
+    // Step 3: insert into the text and where the text flow breaks onto the
+    // next page and using our special character style the page name of the
+    // current page.
+    var win = new Window("palette", "Processing document page numbers...");
+    win.pbar = win.add("progressbar", undefined, 0, tmpDoc.pages.length);
+    win.pbar.preferredSize.width = 300;
+    win.show();
```
@gregoriopellegrino I like this progress bar, thank you! How about we open it when the RTF export begins and update the progress bar with every step — not just for this step 3. That way the user gets feedback for the _entire_ conversion?

Happy to make the change for you to review.

gregoriopellegrino replied:
Good idea. That was part of the page references plugin, it seems to me that it might make sense to expand it to the whole process (I need to study how to do that).

jenstroeger replied:
I thought that we open the progress window with a max value of `8` (because the RTF conversion has eight steps)
```javascript
    win.pbar = win.add("progressbar", undefined, 0, 8);
```
and then update the progress bar with every step
```javascript
    // Step 4: delete empty graphics, and then export the embedded images.
    win.pbar.value = 4;
```
and so on… 🤔

If that `pbar.value` can take float values then we can consider updating it during every single step, too. For example if the document has 100 pages and step 3 iterates over all pages, then with every page we can update
```javascript
    win.pbar.value += (1 / doc.pages.length);
```
which would give a more accurate update with every one of the 8 big steps.

gregoriopellegrino replied:
Looking at Adobe's [JavaScript Tools Guide](https://wwwimages2.adobe.com/content/dam/acom/en/devnet/scripting/pdfs/javascript_tools_guide.pdf) specification, it is not specified if `win.pbar` can be float. I'm under the impression that it's implied to be an `int`.

My proposal is to multiply the steps by 100, then:
```javascript
    win.pbar = win.add("progressbar", undefined, 0, 8 * 100);
```

So you can update in cents (approximating to the nearest integer).

```javascript
    win.pbar.value += Math.round((1 / doc.pages.length) * 100);
```

jenstroeger replied:
Commit https://github.com/bookalope/InDesign-CEP/pull/13/commits/85f5d7b200fcb55bd1e3f1bdad3549c5796167b3.

gregoriopellegrino commented:
### TODO Ignore or handle other link statuses?, line #470

They can be ignored.

gregoriopellegrino commented:
### TODO What about text frames at index 1 and more?, line #478

This part of the script anchors the images to the text on the page (if they are not already manually anchored by the user). I haven't implemented any particular logic (such as anchoring the image to the nearest text box), which seems hyper-complex to me. For now I have used it on my files and it seems to give satisfactory results.

The idea is that each image must be anchored in order to be exported correctly to the RTF. So I anchor all the images to the first text box on the page.

We can tell the user that if they have unexpected results (images anchored in incorrect places) they can anchor them manually before running the script.

gregoriopellegrino commented:
### TODO Revisit, line #493

Repositioning the images for our purpose is not significant (we lose this information in the RTF), so lines 485-496 can be deleted.

jenstroeger committed:
Use progress bar window for the complete RTF conversion.

jenstroeger committed:
Update comments.

jenstroeger commented:
> > TODO doc.links.forEach(), line #250
>
> Fine for me

Looks like [`forEach()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) isn’t supported, gives me an error. So let’s just stick with a simple loop.

> > TODO Consider using a UUID4 filename, line #272
>
> […] The solution that's there now, although "homemade", seems pretty robust to me.

Yes, fair enough. In case you’re curious, the JavaScript client has a `makeUUID4()` function:

https://github.com/bookalope/InDesign-CEP/blob/067bf1947753958a11024d02cd6c9a6b1688e4fe/extensions/Bookalope/js/bookalope.js#L263-L275

> > TODO Notify caller/user that exporting the image failed., line #369
>
> Okay... The thing I find difficult is explaining to the user which image is causing problems. […]

Makes sense.

> > TODO Ignore or handle other link statuses?, line #470
>
> They can be ignored.

Ok.

> > TODO What about text frames at index 1 and more?, line #478
>
> This part of the script anchors the images to the text on the page (if they are not already manually anchored by the user). […]

Thank you for the explanation!

> > TODO Revisit, line #493
>
> Repositioning the images for our purpose is not significant (we lose this information in the RTF), so lines 485-496 can be deleted.

You mean the entire content of this `if` statement can be removed?

https://github.com/bookalope/InDesign-CEP/blob/067bf1947753958a11024d02cd6c9a6b1688e4fe/extensions/Bookalope/jsx/bookalope.jsx#L484-L496

I changed the comments accordingly in commit https://github.com/bookalope/InDesign-CEP/pull/13/commits/b43a6e43b69c7298ecad9421c8f375aa9b4a3ffa.

jenstroeger committed:
Huh, so one-line comments can't contain @param? That's a syntax error.

gregoriopellegrino commented:
> > > TODO doc.links.forEach(), line #250
> > 
> > 
> > Fine for me
> 
> Looks like [`forEach()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) isn’t supported, gives me an error. So let’s just stick with a simple loop.

I think you should use [`everyItem()`](https://www.indesignjs.de/extendscriptAPI/indesign-latest/#Links.html#d1e300372__d1e300445)

> > > TODO Revisit, line #493
> > 
> > 
> > Repositioning the images for our purpose is not significant (we lose this information in the RTF), so lines 485-496 can be deleted.
> 
> You mean the entire content of this `if` statement can be removed?
> 
> https://github.com/bookalope/InDesign-CEP/blob/067bf1947753958a11024d02cd6c9a6b1688e4fe/extensions/Bookalope/jsx/bookalope.jsx#L484-L496
> 
> I changed the comments accordingly in commit [b43a6e4](https://github.com/bookalope/InDesign-CEP/commit/b43a6e43b69c7298ecad9421c8f375aa9b4a3ffa).

Right

jenstroeger commented:
> > You mean the entire content of this `if` statement can be removed?
> > https://github.com/bookalope/InDesign-CEP/blob/067bf1947753958a11024d02cd6c9a6b1688e4fe/extensions/Bookalope/jsx/bookalope.jsx#L484-L496
> > 
> 
> Right

…except the call to `anchorImage()` at the beginning of the `if` block 😉

Also, @gregoriopellegrino, do you develop & test on Windows or on Mac?

jenstroeger committed:
Remove superfluous (debug) code.

jenstroeger commented:
> > …except the call to `anchorImage()` at the beginning of the `if` block 😉
> 
> Yes

Commit https://github.com/bookalope/InDesign-CEP/pull/13/commits/e5ac682f6c1b1e018094ebc1e06cf5eda4841e04.

> > Also, @gregoriopellegrino, do you develop & test on Windows or on Mac?
> 
> Mac

I asked because one tester on Windows gets a [`result.err` of `3`](https://github.com/bookalope/InDesign-CEP/blob/btn-push-active-doc/extensions/Bookalope/js/CEP/CEPEngine_extensions-11.0.0.js#L70-L73) from the extension when the extension panel tries to read the generated RTF:

https://github.com/bookalope/InDesign-CEP/blob/e5ac682f6c1b1e018094ebc1e06cf5eda4841e04/extensions/Bookalope/js/bookalope.js#L595-L599

Unfortunately, [exportFile()](https://www.indesignjs.de/extendscriptAPI/indesign-latest/#Story.html#d1e531497__d1e538209) (on the InDesign side) doesn’t document failure — does it throw an exception, perhaps? Should we set `forceSave=true`?

gregoriopellegrino commented:
Maybe the problem is that it doesn't find the destination path? I would have to do some testing... But I don't have a windows...

jenstroeger committed:
Make the extension work on Windows. Sigh.

jenstroeger commented on extensions/Bookalope/jsx/bookalope.jsx:
```diff
@@ -193,3 +193,382 @@ function bookalopeCreateDocument(idmlFileName, bookId, bookflowId, betaHost) {
+        return 1;
+    }
+
+    // Create a temporary copy of the document that we want to export.
+    const tmpFile = File(app.createTemporaryCopy(doc.fullName))
+    const tmpDoc = app.open(tmpFile, false);
+    const tmpPath = tmpFile.path;
```
On Windows, the [`app.createTemporaryCopy()`](https://www.indesignjs.de/extendscriptAPI/indesign-latest/#Application.html#d1e42253__d1e45703) creates a path _exactly like the original_, whereas on Mac it returns a path in the system’s temp folder. [I asked in Adobe’s Slack channel](https://adobedevs.slack.com/archives/C1FKLQ63F/p1635906342026700), let’s see what comes up. The result of that is that
```javascript
tmpDoc.close(SaveOptions.NO);
```
in [line 571](https://github.com/bookalope/InDesign-CEP/pull/13/files#diff-a7f0d70ffb3265087febcf37d8c4737fbac47f3bf7e9e1acfcc1d3f650407544R571) closes the active, open document. Sigh.

gregoriopellegrino replied:
Wow, that's quite a bug!

jenstroeger replied:
Yes it’s weird and reproducible 😕 I wonder if we missed something, but it _should_ work the same way on both platforms?

Well, let’s see what the community responds in Slack.

gregoriopellegrino replied:
On my mac the script seems to work, but I get `Bookalope server error: Unable to set document because one is already set` from the UI

jenstroeger replied:
Hm, that’s from calling [this endpoint](https://github.com/bookalope/Bookalope/blob/master/API.md#post-bookflows-files-document) here:

https://github.com/bookalope/InDesign-CEP/blob/4daddea57165a28163b6918d6bee743950ab7fd3/extensions/Bookalope/js/bookalope.js#L601-L607

What happens is that the extension uploads a document for a Bookflow that already has a document uploaded. Did you encounter an other error before that, and tried to upload again? Hmm 🤔

jenstroeger replied:
> Well, let’s see what the community responds in Slack.

They asked me to file a bug, and suggested to work around it:

> That said, you should be able to pretty easily cook up your own “TempFileCopy()” (or whatever) command using standard ExtendScript code. These APIs would probably take you 99% of the way:
[File.copy()](https://github.com/bbb999/Types-for-Adobe/blob/c34df41331f7ff9cb4cd92352c8ecb2adbd449bf/shared/JavaScript.d.ts#L1651) (instance method)
[Folder.temp](https://github.com/bbb999/Types-for-Adobe/blob/c34df41331f7ff9cb4cd92352c8ecb2adbd449bf/shared/JavaScript.d.ts#L1854) (static property)

jenstroeger reviewed (commented):
@gregoriopellegrino can you please take a look at commit https://github.com/bookalope/InDesign-CEP/pull/13/commits/668a3b075fb1647974d6e2b68984896ec07f69fd — the commit fixes the extension and makes it work on Windows.

Basically, the CEP function `bookalopeActiveDocumentToRTF()` now produces the filename for the generated RTF file, and it returns that filename to the extension panel; if something fails it returns `false`. So, no more passing a filename to CEP because I could not figure out why that approach doesn’t work 🤦🏻‍♂️

jenstroeger committed:
Remove undefined variable (former function argument).

gregoriopellegrino commented:
Here you have the screen recording of the bug in uploading the document (it is a new document, never uploaded before): https://drive.google.com/file/d/1UOyzoC6n274zay91w6ri8wih6THRdHDN/view?usp=sharing

jenstroeger commented:
> Here you have the screen recording of the bug in uploading the document (it is a new document, never uploaded before): https://drive.google.com/file/d/1UOyzoC6n274zay91w6ri8wih6THRdHDN/view?usp=sharing

Did you build that extension yourself?

With the new workflow I renamed the endpoint and the internal Bookflow steps. That change isn’t part of this PR 😉 The old endpoints forward to the new endpoints, so that clients don’t break but this InDesign extension has one test in `js/uses/BookalopeClient.js:1370` (at the bottom):

```diff
diff --git a/extensions/Bookalope/js/uses/BookalopeClient.js b/extensions/Bookalope/js/uses/BookalopeClient.js
index 46215ac..1e66cad 100644
--- a/extensions/Bookalope/js/uses/BookalopeClient.js
+++ b/extensions/Bookalope/js/uses/BookalopeClient.js
@@ -1246,7 +1246,7 @@ Bookflow.prototype.getImage = function(name) {
   var bookalope = bookflow._bookalope;
 
   return new Promise(function(resolve, reject) {
-    var url = bookflow.url + "/files/image";
+    var url = bookflow.url + "/upload/image";
     var params = {
       "name": name
     };
@@ -1298,7 +1298,7 @@ Bookflow.prototype.addImage = function(name, filename, file) {
     if (bookflow.step !== "convert") {
       reject(new BookalopeError("Unable to add image if Bookflow is not in 'convert' step."));
     } else {
-      var url = bookflow.url + "/files/image";
+      var url = bookflow.url + "/upload/image";
       var params = {
         "file": btoa(file),
         "filename": filename,
@@ -1329,7 +1329,7 @@ Bookflow.prototype.getDocument = function() {
   var bookalope = bookflow._bookalope;
 
   return new Promise(function(resolve, reject) {
-    var url = bookflow.url + "/files/document";
+    var url = bookflow.url + "/upload/document";
     var params = undefined;
     var options = {
       "responseType": "blob"
@@ -1366,10 +1366,10 @@ Bookflow.prototype.setDocument = function(filename, file, filetype, skip_analysi
   var bookalope = bookflow._bookalope;
 
   return new Promise(function(resolve, reject) {
-    if (bookflow.step !== "files") {
+    if (bookflow.step !== "upload") {
       reject(new BookalopeError("Unable to set document because one is already set"));
     } else {
-      var url = bookflow.url + "/files/document";
+      var url = bookflow.url + "/upload/document";
       var params = {
         "file": btoa(file),
         "filename": filename,
```
The 
```javascript
if (bookflow.step !== "files") {
```
line should be `"upload"`.

gregoriopellegrino commented:
I'm using the source from the `btn-push-active-doc` branch

jenstroeger commented:
> I'm using the source from the `btn-push-active-doc` branch

Yes, then you need to change that test

https://github.com/bookalope/InDesign-CEP/blob/4daddea57165a28163b6918d6bee743950ab7fd3/extensions/Bookalope/js/uses/BookalopeClient.js#L1369

to `"upload"`.

jenstroeger committed:
Fix various minor syntax errors, and fix variable declaration.

jenstroeger committed:
Don't shadow variable name `tmpFile` from outer scope.

Considering that `var` declares in function scope, this (I think)
redefines the outer variable. Bug.

jenstroeger committed:
Add required/missing (?) `new` operator.

jenstroeger commented:
@gregoriopellegrino, I’ve added a few minor adjustments (see the three commits above) and also changed this code

https://github.com/bookalope/InDesign-CEP/blob/84db3cee45e973c3c9acb1f0030e1926379636a3/extensions/Bookalope/jsx/bookalope.jsx#L395-L398

to this:
```javascript
    // Create a temporary copy of the document that we want to export.
    const docFile = doc.fullName;  // This is a File object.
    const tmpPath = Folder.temp;
    const tmpFileName = tmpPath + "/" + docFile.name;  // TODO Create a unique non-existent filename?
    if (!docFile.copy(tmpFileName)) {
        alert("Unable to create a temporary copy of the document");
        return JSON.stringify(false);
    }
    const tmpFile = new File(tmpFileName);
    const tmpDoc = app.open(tmpFile, false);
```
Waiting for feedback from the tester on Windows; my local Windows copy has expired for now.

But, there’s also a problem with the sequence of page breaks. Take a look at this screenshot:

![pb](https://user-images.githubusercontent.com/12053937/142717735-0ba3097a-ebe6-4a26-9da8-cb06bd3a5d11.jpg)

The page breaks were extracted from the RTF that this extension pushed to Bookalope… 🤔

gregoriopellegrino commented:
Really strange. Can you share the InDesign file, or the RTF?

jenstroeger commented:
> Really strange. Can you share the InDesign file, or the RTF?

I’ll ask the tester, yes.

jenstroeger committed:
Replace `const` with `var` declarations.

Bit of a tricky decision: ExtendScript3 is based on ECMAScript3 which
does not support `const` and `let` and block scopes. However, apparently
“someone” added `const` support in function scope (like `var`) to have
some basic support for constants.

That’s misleading, coming from more modern JavaScript, and so I decided
to use `var` because that’s what the language supports.

jenstroeger committed:
Remove duplicated code: `isToken()` lives in BookalopeClient.js

jenstroeger committed:
Typo.

jenstroeger committed:
Prefer to be stricter when comparing values.

gregoriopellegrino commented:
Okay, I've partially found the problem.

Page breaks 3 and 4 do not exist in the original InDesign, but are inserted by the script when it joins all the text frames.

However, this character is not a page break, but a frame break (at least that's how it is inside InDesign). I can't figure out how this separation character is exported to the RTF.

Eventually in the script I can insert a simple carriage return, so it shouldn't cause us any problems, but can you inspect the RTF first?

jenstroeger commented:
> Eventually in the script I can insert a simple carriage return, so it shouldn't cause us any problems, …

The carriage return would be ingested into Bookalope as a soft linebreak, that would be ok.

> … but can you inspect the RTF first?

You mean some form of preprocessing before Bookalope works through the RTF? What would you want to check during preprocessing?

gregoriopellegrino commented:
> > … but can you inspect the RTF first?
> 
> You mean some form of preprocessing before Bookalope works through the RTF? What would you want to check during preprocessing?

My question is: can you check why the import script interprets the frame break character as a page break?

jenstroeger commented:
> My question is: can you check why the import script interprets the frame break character as a page break?

When Bookalope ingests the RTF and the RTF text contains a number that is formatted with the `"bookalope-page-number"` character style then Bookalope ingests a page break with that number:

https://github.com/bookalope/InDesign-CEP/blob/92cd5ba64b2d21f30b2e45356e440522489d38f6/extensions/Bookalope/jsx/bookalope.jsx#L417-L427

So, here

https://github.com/bookalope/InDesign-CEP/blob/92cd5ba64b2d21f30b2e45356e440522489d38f6/extensions/Bookalope/jsx/bookalope.jsx#L450-L451

I wonder if a frame break is also included? But if so, why are the numbers out of order? Unfortunately, I’m not too familiar with InDesign’s workings 😕

gregoriopellegrino commented:
![142717735-0ba3097a-ebe6-4a26-9da8-cb06bd3a5d11](https://user-images.githubusercontent.com/5065125/143639067-750db418-5ba4-4a25-85dc-d9f466ca0079.jpeg)

From my analysis:

- it seems to me that the system defines two parallel page-number sequences
- I have identified the red one (which in my opinion is based on the frame break character) and the blue one which is based on the `bookalope-page-number` character style
- I have not been able to understand why in the blue numbering there is the `5`, this I will investigate after having "resolved" the issue of the break frame character

So my question is, can the system distinguish the frame break from the page break? If not, I'll modify the script...

gregoriopellegrino commented:
For the sake of completeness, the frame break character is inserted here:

https://github.com/bookalope/InDesign-CEP/blob/92cd5ba64b2d21f30b2e45356e440522489d38f6/extensions/Bookalope/jsx/bookalope.jsx#L560-L561

jenstroeger commented:
> So my question is, can the system distinguish the frame break from the page break? If not, I'll modify the script...

Hmm, good question: how does that `FRAME_BREAK` translate into RTF? I’ll take a look at the Bookalope side. You want the frame break to become a simple soft linebreak, correct? Or ignored?

gregoriopellegrino commented:
> Hmm, good question: how does that `FRAME_BREAK` translate into RTF? I’ll take a look at the Bookalope side. You want the frame break to become a simple soft linebreak, correct? Or ignored?

Turning it into a paragraph break I think would be the best thing: this character is only inserted if the frames are not linked, so it can't be that in InDesign there is a paragraph that is common to two unlinked frames.

jenstroeger commented:
Oh 😳 I think this is my fault: the RTF is a paged document and it contains the InDesign page breaks (marked up using the `"bookalope-page-number"` character style) plus the actual page breaks when the RTF is rendered onto a page. **Bookalope ingests both.**

In the [above screenshot](https://user-images.githubusercontent.com/12053937/142717735-0ba3097a-ebe6-4a26-9da8-cb06bd3a5d11.jpg) page breaks `3`, `4`, `5` are the actual document page breaks, and `1`, `2` are the InDesign breaks inserted into the RTF. We probably want to ignore the actual document breaks and extract _only_ the ones that InDesign inserted?

gregoriopellegrino commented:
> In the [above screenshot](https://user-images.githubusercontent.com/12053937/142717735-0ba3097a-ebe6-4a26-9da8-cb06bd3a5d11.jpg) page breaks `3`, `4`, `5` are the actual document page breaks, and `1`, `2` are the InDesign breaks inserted into the RTF. We probably want to ignore the actual document breaks and extract _only_ the ones that InDesign inserted?

Yes

> So… if we could add _all_ InDesign page information into the RTF then Bookalope could ignore the actual rendered page breaks and ingest _only_ the InDesign page break information.

The script will insert the page information if there is a text frame on the page. If the page is empty, then there will be no number.... That's correct in my opinion, what do you think?

jenstroeger commented:
I agree that it’s correct, though perhaps incomplete looking at it from a “digital first” viewpoint: ideally we’d like to enumerate _all_ print pages into the page list.

Perhaps we should use the InDesign page breaks not as an absolute number but as a marker to increment? If we assume that the RTF contains forced page breaks then the markers could continue to increment based on the InDesign markers — ignoring the flow breaks that occur through rendering by the app… Hmm…

gregoriopellegrino commented:
From an accessibility point of view these are the requirements: https://www.w3.org/TR/epub-a11y-11/#sec-page-list

Eventually blank pages can be listed as displayed here: https://github.com/w3c/epub-specs/issues/1502#issue-800500840

jenstroeger commented:
Let me explore the RTF format some more: it has explicit `\page` and `\pagebb` control words that _require_ a page break when rendered. That, together with the InDesign markers might be enough to reconstruct a page list<sup>*</sup> — maybe not complete but hopefully almost 🤞🏼

—————
<sup>*</sup> Let’s ignore front/main/back matter numbering for now.

jenstroeger commented:
@gregoriopellegrino, the example RTF above contains two forced page breaks (i.e. `\page` control words) which are recognized correctly.

If I turn off page-break ingestion (the rendered page breaks that are introduced by flowing the RTF across pages) then we get a document that contains _only_ the InDesign page breaks (this is after a slight rearranging of paragraphs) —

![rtf](https://user-images.githubusercontent.com/12053937/145160097-809d676d-6fa3-48b7-9268-929fac8a26aa.jpg)

So perhaps it’s a good idea to _turn off_ page-break import for RTF, although Bookalope won’t know whether an RTF arrives from InDesign or from a user. Or ignore page breaks if InDesign page breaks exist in the document? 🤔

gregoriopellegrino commented:
> If I turn off page-break ingestion (the rendered page breaks that are introduced by flowing the RTF across pages) then we get a document that contains _only_ the InDesign page breaks (this is after a slight rearranging of paragraphs) —

Great!

> So perhaps it’s a good idea to _turn off_ page-break import for RTF, although Bookalope won’t know whether an RTF arrives from InDesign or from a user. Or ignore page breaks if InDesign page breaks exist in the document? 🤔

The InDesign RTF can only arrive from the InDesign extension. Maybe in the creation of the bookflow we can define a flag or similar to tell the system to ignore page-breaks. There is (or was) a flag for that in the web interface, isn't it?

> Regarding the code change to generating a temporary copy of the active document (mentioned [above](https://github.com/bookalope/InDesign-CEP/pull/13#issuecomment-974606202)) — that change works on Windows for me. Huzza!

Hurra!

jenstroeger committed:
Create a tmp copy of the active document manually, to make Windows work.

jenstroeger committed:
Don’t read pagebreaks when pushing the active document to the server.

jenstroeger commented:
> The InDesign RTF can only arrive from the InDesign extension. Maybe in the creation of the bookflow we can define a flag or similar to tell the system to ignore page-breaks. There is (or was) a flag for that in the web interface, isn't it?

See commit https://github.com/bookalope/InDesign-CEP/pull/13/commits/3538aef961afa74efe9e9617e0f96b47029e7ce7: when the extension uploads the RTF for the active InDesign document, it explicitly disables pagebreaks. I had to make a few changes to the server API for this, but the feature makes a lot of sense. These changes will roll out to Beta next week.

gregoriopellegrino commented:
Great!

jenstroeger committed:
Update Bookalope token check to meet the according server changes.

jenstroeger reviewed (dismissed):
@nGolubeva and @gregoriopellegrino, it’s time to merge, I think.

gregoriopellegrino commented:
Great!

Il giorno sab 22 gen 2022 alle 22:21 Jens Tröger ***@***.***>
ha scritto:

> ***@***.**** approved this pull request.
>
> @nGolubeva <https://github.com/nGolubeva> and @gregoriopellegrino
> <https://github.com/gregoriopellegrino>, it’s time to merge, I think.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bookalope/InDesign-CEP/pull/13#pullrequestreview-860242642>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGUTJJ3Q2V7ULSFWIE5OWDUXMNVTANCNFSM44DS5Y2A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Gregorio Pellegrino

Effatà Editrice
T 0121 353452
F 0121 353839
E ***@***.***
W www.effata.it/
TW @g_pellegrino <http://twitter.com/g_pellegrino>
FB facebook.com/EffataEditrice
ME about.me/gregorio.pellegrino
IN linkedin.com/in/gregoriopellegrino
<https://it.linkedin.com/in/gregoriopellegrino>

jenstroeger committed:
Update README.
